snapshotter/snapshotter/layer.go

```diff
@@ -0,0 +1,292 @@
+			DefaultDateString, DateFormat, err)
+	}
+
+	// Create tarfile for writing
+	tarFile, err := os.Create(tarPath)
+	if err != nil {
+		return fmt.Errorf("failed to create tar file %s. %w", tarPath, err)
+	}
+	defer tarFile.Close()
+
+	// Create a tar writer.
+	w := tar.NewWriter(tarFile)
+
+	// Use WalkDir to iterate through the contents of the directory.
+	// WalkDir guarantees lexicographical ordering.
+	err = filepath.WalkDir(dirPath, func(path string, entry fs.DirEntry, err error) error {
+		// Return early if there is an error
+		if err != nil {
+			return err
+		}
+
+		// Fetch the FileInfo from the entry.
+		info, err := entry.Info()
+		if err != nil {
+			return fmt.Errorf("failed to fetch Info from %v. %w", entry, err)
+		}
+
+		var header *tar.Header
+
+		// If the file is a symlink, fetch the target.
+		if (info.Mode() & fs.ModeSymlink) != 0 {
+			link, err := os.Readlink(path)
+			if err != nil {
+				return fmt.Errorf("failed to readlink %s. %w", path, err)
+			}
+
+			// Create header will set the target field to link.
+			header, err = tar.FileInfoHeader(info, link)
+		} else {
+			// Create header.
+			header, err = tar.FileInfoHeader(info, info.Name())
+		}
+
+		if err != nil {
+			return fmt.Errorf("failed to create FileInfoHeader. %w", err)
+		}
+
+		// Remove the prefix (dirPath) from each path.
+		relativePath := path[len(dirPath):]
+		header.Name = relativePath
+
+		if path == dirPath {
+			// For root folder, set the modification time to default time.
+			header.Name = "/"
+			header.ModTime = defaultDate
+		}
+
+		// Set the access and change times to the modification time.
+		// This is essential for reproducibility.
+		// See: https://reproducible-builds.org/docs/archives/
+		header.AccessTime = header.ModTime
+		header.ChangeTime = header.ModTime
+
+		// Write the header
+		if err := w.WriteHeader(header); err != nil {
+			return fmt.Errorf("failed to write header. %w", err)
+		}
+
+		// Skip writing contents if not a regular file.
+		if !info.Mode().IsRegular() {
+			return nil
+		}
+
+		// Write the contents of the file to the tar archive.
+		file, err := os.Open(path)
+		if err != nil {
+			return fmt.Errorf("failed to open file %s for copying. %w",
+				path, err)
+		}
+		defer file.Close()
+
+		_, err = io.Copy(w, file)
+		if err != nil {
+			return fmt.Errorf("failed to copy %s. %w", path, err)
+		}
+		return err
+	})
+	return err
+}
+
+// Compute md5 hash of given data
+func computeMD5Hash(data []byte) [16]byte {
+	var sum [16]byte
+	hash := md5.Sum(data)
+	// Retype to [16]byte
+	copy(sum[:], hash[:16])
+	return sum
+}
+
+// Compute the dmverity merkeltree and append to the disk image file.
+func computeAndWriteHashDevice(diskImagePath string) (string, error) {
+	w, err := os.OpenFile(diskImagePath, os.O_RDWR, 0644)
+	if err != nil {
+		return "", fmt.Errorf("failed to open %s. %w", diskImagePath, err)
+	}
+
+	// Find the size of the disk image.
+	ext4blocksize := 4096
+	ext4size, err := w.Seek(0, io.SeekEnd)
```

Comment:
Does this seek to the end of the disk image? Is the disk image padded out to 4096 bytes?